lint/long-chain_checker.go

```diff
@@ -0,0 +1,121 @@
+package lint
+
+import (
+	"fmt"
+	"go/ast"
+	"strings"
+
+	"github.com/Quasilyte/astcmp"
+)
+
+type longChainChecker struct {
+	ctx *Context
+}
+
+func newLongChainChecker(ctx *Context) Checker {
+	return &longChainChecker{
+		ctx: ctx,
+	}
+}
+
+func (c *longChainChecker) Check(f *ast.File) {
+	for _, decl := range f.Decls {
+		if decl, ok := decl.(*ast.FuncDecl); ok {
+			if decl.Body == nil {
+				continue
+			}
+			for _, stmt := range decl.Body.List {
+				if stmt, ok := stmt.(*ast.SwitchStmt); ok {
+					c.checkSwitch(stmt)
+				}
+			}
+		}
+	}
+}
+
+func (c *longChainChecker) checkSwitch(stmt *ast.SwitchStmt) {
+	exprs := []ast.Expr{}
+	for _, st := range stmt.Body.List {
+		s := st.(*ast.CaseClause)
+		if s.List == nil {
+			continue
+		}
+		exprs = append(exprs, s.List...)
+	}
+	if len(exprs) < 2 {
+		return
+	}
+
+	cp := c.exprToList(exprs[0])
+	for _, e := range exprs[1:] {
+		cp = c.commonPrefix(cp, c.exprToList(e))
+	}
+
+	const n = 3
+
+	if len(cp) > n {
+		c.warn(cp, stmt)
+	}
+}
+
+func (c *longChainChecker) exprToList(expr ast.Expr) []ast.Expr {
+	res := []ast.Expr{}
+	tmp := expr
+	for {
+		switch t := tmp.(type) {
+		case *ast.SelectorExpr:
+			res = append(res, t.Sel)
+			tmp = t.X
+		default:
+			res = append(res, tmp)
+			res = reverseExprSlice(res)
+			return res
+
+		}
+	}
+}
+
+func reverseExprSlice(e []ast.Expr) []ast.Expr {
+	for i := len(e)/2 - 1; i >= 0; i-- {
+		j := len(e) - 1 - i
+		e[i], e[j] = e[j], e[i]
+	}
+	return e
+}
+
+func (c *longChainChecker) commonPrefix(a, b []ast.Expr) []ast.Expr {
```

Comment:
It's better to call these `x` and `y`.
Even better, `xs` and `ys`.

Reply:
ok